src/server/server_family.cc

```diff
@@ -1993,79 +1993,102 @@ GenericError ServerFamily::DoSaveCheckAndStart(const SaveCmdOptions& save_cmd_op
                           "SAVING - can not save database"};
     }
-  }
 
-  // Create save controller outside of mutex to avoid blocking INFO commands
-  auto snapshot_storage = save_cmd_opts.cloud_uri.empty()
-                              ? snapshot_storage_
-                              : CreateCloudSnapshotStorage(save_cmd_opts.cloud_uri);
+    auto snapshot_storage = save_cmd_opts.cloud_uri.empty()
+                                ? snapshot_storage_
+                                : CreateCloudSnapshotStorage(save_cmd_opts.cloud_uri);
 
-  auto temp_save_controller = make_unique<SaveStagesController>(detail::SaveStagesInputs{
-      save_cmd_opts.new_version, save_cmd_opts.cloud_uri, save_cmd_opts.basename, trans, &service_,
-      fq_threadpool_.get(), snapshot_storage, opts.bg_save});
+    controller = make_shared<SaveStagesController>(detail::SaveStagesInputs{
+        save_cmd_opts.new_version, save_cmd_opts.cloud_uri, save_cmd_opts.basename, trans,
+        &service_, fq_threadpool_.get(), snapshot_storage, opts.bg_save});
+    save_controller_ = controller;
+  }
 
   // Initialize resources outside of mutex (this may take time for S3 operations)
-  auto res = temp_save_controller->Init();
+  auto res = controller->Init();
+  if (res) {
+    DCHECK_EQ(res->error, true);
+    thread_safe_save_info_.Update([&](SaveInfoData* data) {
+      data->last_error = res->error;
+      data->last_error_time = res->save_time;
+      data->failed_duration_sec = res->duration_sec;
+      if (bg_save) {
+        data->last_bgsave_status = false;
+      }
+    });
 
-  // Now acquire mutex only to set the controller and update state
-  {
+    // Reset the controller under lock if initialization failed.
     util::fb2::LockGuard lk(save_mu_);
-
-    // Double-check that no other save started while we were initializing
-    if (save_controller_) {
-      return GenericError{make_error_code(errc::operation_in_progress),
-                          "SAVING - can not save database"};
+    if (save_controller_ == controller) {
+      save_controller_.reset();
     }
+    return res->error;
+  }
 
-    if (res) {
-      DCHECK_EQ(res->error, true);
-      last_save_info_.SetLastSaveError(*res);
-      // Don't set save_controller_ since initialization failed
-      if (bg_save) {
-        last_save_info_.last_bgsave_status = false;
-      }
-      return res->error;
-    }
+  // Success - update state
+  controller->Start();
+  thread_safe_save_info_.Update(
+      [bg_save](SaveInfoData* data) { data->bgsave_in_progress = bg_save; });
 
-    // Success - set the controller and update state
-    save_controller_ = std::move(temp_save_controller);
-    save_controller_->Start();
-    last_save_info_.bgsave_in_progress = bg_save;
-  }
   return {};
 }
 
 GenericError ServerFamily::WaitUntilSaveFinished(Transaction* trans, bool ignore_state) {
-  save_controller_->WaitAllSnapshots();
+  std::shared_ptr<SaveStagesController> controller;
```

Comment:
Haven't looked into this part of town for years - it looks like the previous code relied on the fact that WaitUntilSaveFinished was always called only when the before DoSaveCheckAndStart succeeded. So theoretically `shared_ptr` isn't needed anywhere, but without it also looks to fragile